### PR TITLE
Replaced insecure random with proper urandom

### PR DESCRIPTION
**Pull request by yackermann**
Increased length of the password to 16

Replaced random with urandom

Fixed shebang to be more portable

### CODE REVIEW

**Comment by joshua-d-miller** [comment]
So I'm confused about increasing the password length because what I have set is a default setting. You can change the length with an MDM or by setting the value of PasswordLength in the PLIST file. What are the advantages to **urandom** vs the standard **randint**?

File: macOSLAPS [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
```
Comment (clburlison) [inline macOSLAPS]: I don't believe making this script more 'portable' is the correct way to go. In fact this change breaks this script if a user has python3 set as their main Python Interpreter or [one of many other ways to break things](http://stackoverflow.com/a/18043508/4811765) #thankshomebrew. This script is already targeted for macOS: uses `dscl` and `CFPreferencesCopyAppValue` from the Foundation module. I'd say you should revert this. With the introduction of SIP in 10.11 it is much more likely that `#!/usr/bin/python` will function properly across a wider range of machines.

Comment (yackermann) [inline macOSLAPS]: @clburlison Well, current shebang will still cause an issue with python3 beeing default. Possible solution

``` bash
#!/usr/bin/env python2.7
```

P.S. I don't see an issue with this script running on python3.

**Comment by yackermann** [comment]
@joshua-d-miller standard python random, is a Linear Congruential Generator. It is simply 

```
new_random_number = (prev_random_number * some_large_number) % some_large_prime
```

As you can see here, at the start of generation, the initial will be always the same, as `prev_random_number` is not set, and tend to be set to 1. To avoid it, libraries use time as a seed for initial state of the generator.

This is terribly insecure, and give max of 2^32 entropy, and makes it no issue to simple bruteforce your password in no time.

To solve this issue we use `urandom`, which is python library that utilises OS built in CSPRNG (Cryptographically secure pseudorandom number generator). This generator picks up random events, such as disk read operations, user clicks, bus event, and mixes them together, generating very secure random numbers with high entropy.  Filippo Valsorda gave a nice talk on urandom https://media.ccc.de/v/32c3-7441-the_plain_simple_reality_of_entropy

**Comment by yackermann** [comment]
@joshua-d-miller The reason to increase password length was that I was not sure if AD it would support base64, and base16 would give not enough entropy. But it looks like AD supports base 64 https://technet.microsoft.com/en-us/magazine/ff848710.aspx

**Comment by yackermann** [comment]
@joshua-d-miller from python documentation:

> Almost all module functions depend on the basic function random(), which generates a random float uniformly in the semi-open range [0.0, 1.0). Python uses the Mersenne Twister as the core generator. It produces 53-bit precision floats and has a period of 2**19937-1. The underlying implementation in C is both fast and threadsafe. The Mersenne Twister is one of the most extensively tested random number generators in existence. However, being completely deterministic, it is not suitable for all purposes, and is completely unsuitable for cryptographic purposes.

https://docs.python.org/3.5/library/random.html
https://docs.python.org/2.7/library/random.html

**Comment by joshua-d-miller** [comment]
Ok. This makes sense to me and I was messing with the urandom function. I have my PasswordLength set to 14 but it generates a 20 character password. Not sure what that is.

**Comment by yackermann** [comment]
@joshua-d-miller Base64 is 2^6, so 6 bit per char. They are encoded in three octets(3 bytes == 4 chars).  You asked urandom for 14 bytes, which is 4 octets and two bytes. So to fill last octet we add `==` which basically says to decoder "We have two chars that are not complete".

https://en.wikipedia.org/wiki/Base64

**Comment by yackermann** [comment]
@joshua-d-miller Fixed it. Keep the password size 12 min

**Comment by joshua-d-miller** [comment]
So I basically rewrote a lot of the code and I'm importing random. According to Mike Lynn, random should be fine so this may no longer be needed however I'm still open to _urandom_. Please test new code and let me know the results.

**Comment by pudquick** [comment]
I am the Mike Lynn mentioned here.

What I said needs clarification: The new code uses random.choice

If @herrjemand had made their quote of the documentation on the random module complete, they would have seen this:

> random.seed([x])
> 
> Initialize the basic random number generator. Optional argument x can be any hashable object. If x is omitted or None, current system time is used; current system time is also used to initialize the generator when the module is first imported. If randomness sources are provided by the operating system, they are used instead of the system time (see the os.urandom() function for details on availability).

The key part is the last statement - that importing the random module initializes the seed and operating systems that provide randomness sources (which includes OS X, in this instance), os.urandom is indeed used.

True, now we've moved from using urandom N-length times to select each character of the password to using it once to seed Mersenne Twister - but I'd love for a technical statistical writeup showing how much of a security improvement this provides vs. the clarity of the current (as of writing this) code.

Since it's no longer pinned to when the time of the password change was but is now seeded from os.urandom, I'll make the argument that you'd effectively be splitting hairs between (heat death of the universe) and (heat death of the universe + 1) days of computation at this point in trade for less pythonic code.

**Comment by yackermann** [comment]
@pudquick Please refer to 

> However, being completely deterministic, it is not suitable for all purposes, and is completely unsuitable for cryptographic purposes.

FYI. Don't do crypto your self. Using `urandom` is a standard practice in the industry, and yours suggestions of mixing `random` and `urandom` no just ineffective, but in the long term toxic. 

Talking about safety of current passwords, they close to theoretical 2^6^length

**Comment by pudquick** [comment]
@herrjemand  - You're confusing "random" the module and `random` the OS function.

random.choice is seeded from the OS `urandom` function, as I pointed out, in the documentation.

**Comment by yackermann** [comment]
@pudquick I do know that. I do have an issues with it:
- Since `random.choice` relies on `twister`, what if has security flaws?
- Code readability: right now it simply generates random data, converts to base64 and cuts the excess. Can you even understand what this code does?

``` python
characters = ascii_letters + punctuation + digits
password = "".join(choice(characters)  for x in range(randint(pass_length, pass_length)))
```
- Back to security: What if `ascii_letters + punctuation + digits` compromised?

Lets just keep it simple, and do it right.

**Comment by pudquick** [comment]
@herrjemand  - When I said `the clarity of the current (as of writing this) code` you obviously have not looked at what's been committed since, because it's not what you pasted

https://github.com/joshua-d-miller/macOSLAPS/blob/e04bd461481d802cc08ca8be7de2bee11e8cd4c3/macOSLAPS#L116-L120

"Keep it simple" is exactly what my suggestion (as currently committed) is.

**Comment by yackermann** [comment]
@pudquick nice and horrible. Remove my comment on readability and there two more points

**Comment by joshua-d-miller** [comment]
@herrjemand Have you run this code against an Active Directory? The password is written to an attribute in AD that is in clear text but marked confidential. Only those personnel specified can see that password through the LAPS Utility or ASDI Edit (Domain Admins). I can understand the security concerns but the point of LAPS is the password is different on every machine and if one machine is compromised your fleet is not. Plus if compromised you can set a past expiration date in AD and the script will set a new password.

Also have you tested the newer code that I pushed today?

**Comment by pudquick** [comment]
Horrible? The code that was committed wasn't even my original suggestion, which was:

```
password = ''
for i in range(length):
    password += random.choice(characters)
```

But that triggered pylint about not using the `i` variable. It was either something like this or disabling the pylint warning. I didn't make the final choice.

That being said, I'll bow out of the rest of this conversation.

**Comment by yackermann** [comment]
@joshua-d-miller That's not the point. 

I am arguing that we need to do crypto right. Don't need to invent your own ways of generating random passwords. Just do it right.

**Pull request merged**